File: frontend/src/types/billing.ts
export type BillType = 'OPD' | 'IPD' | 'Lab' | 'Pharmacy' | 'Procedure' | 'Other';

export type PaymentStatus = 'Paid' | 'Partially Paid' | 'Pending' | 'Overdue' | 'Refunded' | 'Cancelled';

export type PaymentMode = 'Cash' | 'UPI' | 'Card' | 'Bank Transfer' | 'Other';

export type ApprovalStatus = 'Pending' | 'Requested' | 'Approved' | 'Rejected' | 'Processed' | 'Cancelled';

export interface BillItem {
  id?: string;
  service_name: string;
  category?: string;
  description?: string;
  quantity: number;
  unit_price: number;
  gross_amount: number;
  discount: number;
  tax: number;
  net_amount: number;
}

export interface Bill {
  id: string;
  bill_number: string;
  patient_id?: string;
  patient_name: string;
  uhid: string;
  appointment_id?: string;
  ipd_number?: string;
  bill_type: BillType;
  department?: string;
  doctor_name?: string;
  gross_amount: number;
  discount_amount: number;
  tax_amount: number;
  net_amount: number;
  paid_amount: number;
  pending_amount: number;
  payment_mode: PaymentMode | string;
  payment_status: PaymentStatus;
  discharge_status?: 'Advance' | 'Interim' | 'Final Discharge';
  bill_date: string;
  due_date?: string;
  billing_staff: string;
  branch: string;
  notes?: string;
  items: BillItem[];
}

export interface PaymentCollection {
  id?: string;
  receipt_number: string;
  bill_id?: string;
  bill_number: string;
  patient_name: string;
  uhid: string;
  service_type: string;
  total_bill: number;
  previously_paid: number;
  current_payment: number;
  remaining_due: number;
  payment_mode: PaymentMode;
  transaction_ref?: string;
  payment_date: string;
  collected_by: string;
  branch: string;
  notes?: string;
}

export interface DiscountRequest {
  id?: string;
  discount_code: string;
  bill_number: string;
  patient_name: string;
  uhid: string;
  original_amount: number;
  discount_type: 'Percentage' | 'Fixed Amount';
  discount_value: number;
  discount_amount: number;
  reason: string;
  requested_by: string;
  approved_by?: string;
  status: ApprovalStatus;
  request_date: string;
}

export interface RefundRequest {
  id?: string;
  refund_code: string;
  bill_number: string;
  patient_name: string;
  uhid: string;
  original_amount: number;
  paid_amount: number;
  refund_amount: number;
  refund_reason: string;
  refund_mode: PaymentMode;
  requested_by: string;
  approved_by?: string;
  refund_date: string;
  status: ApprovalStatus;
}

export interface BillCancellation {
  id?: string;
  cancellation_code: string;
  bill_number: string;
  patient_name: string;
  uhid: string;
  original_amount: number;
  cancellation_reason: string;
  requested_by: string;
  approved_by?: string;
  cancellation_date: string;
  status: ApprovalStatus;
}

export interface SupplierPayable {
  id?: string;
  supplier_name: string;
  invoice_number: string;
  purchase_date: string;
  invoice_amount: number;
  paid_amount: number;
  outstanding_amount: number;
  due_date?: string;
  payment_status: PaymentStatus;
  module_source: 'Pharmacy' | 'Store';
  branch: string;
}

export interface BillingAuditLog {
  id?: string;
  transaction_id: string;
  bill_number?: string;
  entity_type: 'Bill' | 'Payment' | 'Discount' | 'Refund' | 'Cancellation';
  action: 'Created' | 'Updated' | 'Approved' | 'Rejected' | 'Refunded' | 'Cancelled' | 'Collected' | 'Requested';
  previous_value?: string;
  new_value?: string;
  user_name: string;
  user_role: string;
  timestamp: string;
  reason?: string;
}

export interface BillingKPIs {
  today_revenue: number;
  today_billing: number;
  today_collection: number;
  total_outstanding: number;
  today_refunds: number;
  today_discounts: number;
  opd_revenue: number;
  ipd_revenue: number;
  lab_revenue: number;
  pharmacy_revenue: number;
  procedure_revenue: number;
  total_revenue: number;
  total_expenses: number;
  net_revenue: number;
}

export interface CashierSummary {
  opening_cash: number;
  cash_collection: number;
  upi_collection: number;
  card_collection: number;
  bank_collection: number;
  other_collection: number;
  total_collection: number;
  refunds: number;
  net_cash: number;
  closing_cash: number;
  transaction_count: number;
}

File: backend/app/schemas/billing.py
from pydantic import BaseModel, ConfigDict
from typing import Optional, List


class BillItemSchema(BaseModel):
    id: Optional[str] = None
    service_name: str
    category: Optional[str] = "General"
    description: Optional[str] = None
    quantity: int = 1
    unit_price: float = 0.0
    gross_amount: float = 0.0
    discount: float = 0.0
    tax: float = 0.0
    net_amount: float = 0.0

    model_config = ConfigDict(from_attributes=True)


class BillCreateSchema(BaseModel):
    bill_number: Optional[str] = None
    patient_id: Optional[str] = None
    patient_name: str
    uhid: str
    appointment_id: Optional[str] = None
    ipd_number: Optional[str] = None
    bill_type: str  # OPD, IPD, Lab, Pharmacy, Procedure, Other
    department: Optional[str] = None
    doctor_name: Optional[str] = None
    gross_amount: float = 0.0
    discount_amount: float = 0.0
    tax_amount: float = 0.0
    net_amount: float = 0.0
    paid_amount: float = 0.0
    pending_amount: float = 0.0
    payment_mode: str = "Cash"
    payment_status: str = "Pending"
    discharge_status: Optional[str] = None
    bill_date: str
    due_date: Optional[str] = None
    billing_staff: str = "System Billing"
    branch: str = "Main Branch"
    notes: Optional[str] = None
    items: List[BillItemSchema] = []

    model_config = ConfigDict(from_attributes=True)


class BillResponseSchema(BillCreateSchema):
    id: str
    items: List[BillItemSchema] = []

    model_config = ConfigDict(from_attributes=True)


class PaymentCollectionCreateSchema(BaseModel):
    receipt_number: Optional[str] = None
    bill_id: Optional[str] = None
    bill_number: str
    patient_name: str
    uhid: str
    service_type: str = "OPD"
    total_bill: float = 0.0
    previously_paid: float = 0.0
    current_payment: float = 0.0
    remaining_due: float = 0.0
    payment_mode: str  # Cash, UPI, Card, Bank Transfer, Other
    transaction_ref: Optional[str] = None
    payment_date: str
    collected_by: str
    branch: str = "Main Branch"
    notes: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class DiscountRequestSchema(BaseModel):
    id: Optional[str] = None
    discount_code: Optional[str] = None
    bill_number: str
    patient_name: str
    uhid: str
    original_amount: float
    discount_type: str  # Percentage | Fixed Amount
    discount_value: float
    discount_amount: float
    reason: str
    requested_by: str
    approved_by: Optional[str] = None
    status: str = "Pending"
    request_date: str

    model_config = ConfigDict(from_attributes=True)


class RefundRequestSchema(BaseModel):
    id: Optional[str] = None
    refund_code: Optional[str] = None
    bill_number: str
    patient_name: str
    uhid: str
    original_amount: float
    paid_amount: float
    refund_amount: float
    refund_reason: str
    refund_mode: str = "Cash"
    requested_by: str
    approved_by: Optional[str] = None
    refund_date: str
    status: str = "Requested"

    model_config = ConfigDict(from_attributes=True)


class BillCancellationSchema(BaseModel):
    id: Optional[str] = None
    cancellation_code: Optional[str] = None
    bill_number: str
    patient_name: str
    uhid: str
    original_amount: float
    cancellation_reason: str
    requested_by: str
    approved_by: Optional[str] = None
    cancellation_date: str
    status: str = "Pending"

    model_config = ConfigDict(from_attributes=True)


class SupplierPayableSchema(BaseModel):
    id: Optional[str] = None
    supplier_name: str
    invoice_number: str
    purchase_date: str
    invoice_amount: float
    paid_amount: float
    outstanding_amount: float
    due_date: Optional[str] = None
    payment_status: str = "Pending"
    module_source: str = "Pharmacy"
    branch: str = "Main Branch"

    model_config = ConfigDict(from_attributes=True)


class BillingAuditLogSchema(BaseModel):
    id: Optional[str] = None
    transaction_id: str
    bill_number: Optional[str] = None
    entity_type: str
    action: str
    previous_value: Optional[str] = None
    new_value: Optional[str] = None
    user_name: str
    user_role: str
    timestamp: str
    reason: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)
